### DISCUSSION TITLE
Next/Head encode URI?

### DISCUSSION THREAD

**Discussion by gbreux** [vercel/next.js#12480] [Help] [answered]
Hi,

I'd like to understand why inserting a meta inside a `<Head>` will result in an altered content (encodeURI)?

```
<Head>
  <meta property="og:image" content="https://monsite.com/image?prop1=1&prop2=2">
</Head>
```

Will be converted as: (notice the **`&amp;`** before `prop2`)
```
<head>
  <meta property="og:image" content="https://monsite.com/image?prop1=1&amp;prop2=2">
</head>
```

And this will break a netlify redirection (for example)

So why is it that way? Is it a bug?

Thanks for you help!

**Comment by lukeshumard** [comment]
A solution for this is to use an API route that will redirect to the image. This approach encodes the URL for the image, which is then decoded by the API route. The API route will then redirect to your image URL.

```jsx
// pages/index.js
const imageUrl = 'https://monsite.com/image?prop1=1&prop2=2';

<Head>
  <meta property="og:image" content={`https://YOUR_DOMAIN/api/social/${encodeURIComponent(imageUrl)}`} />
</Head>
```

```ts
// pages/api/[url].ts
import { NextApiRequest, NextApiResponse } from 'next';

async function socialImageRedirect(req: NextApiRequest, res: NextApiResponse) {
  const { url } = req.query;
  const redirectUrl = decodeURIComponent(url as string);
  res.redirect(redirectUrl);
}

export default socialImageRedirect;
```

**Comment by mdawaffe** [comment]
(I randomly saw a link to this discussion on Twitter today.)

This looks like the correct behavior to me.

```html
<meta property="og:image" content="https://monsite.com/image?prop1=1&amp;prop2=2">
```

Is the normal HTML that represents a `<meta>` tag with a `content` value of `https://monsite.com/image?prop1=1&prop2=2`.

As a concrete example, check out the source of https://github.com/vercel. As I write this comment, the source shows:

```html
<meta property="og:image" content="https://avatars.githubusercontent.com/u/14985020?s=280&amp;v=4" />
```

Now, on https://github.com/vercel, open your browser's console and do something like:
```js
Object.fromEntries(
	Array.from( document.querySelectorAll('meta') )
		.filter( m => m.getAttribute( 'property' ) )
		.map( m => [ m.getAttribute( 'property' ), m.content ] )
)
```

You'll see that the `og:image`'s `content` property's value really is what you think it should be:
```
{
	"fb:app_id": "1401488693436528",
	"og:image": "https://avatars.githubusercontent.com/u/14985020?s=280&v=4",
	"og:site_name": "GitHub",
	"og:type": "profile",
	"og:title": "Vercel",
	"og:url": "https://github.com/vercel",
	"og:description": "Develop. Preview. Ship. Creators of Next.js. Vercel has 137 repositories available. Follow their code on GitHub.",
	"profile:username": "vercel"
}
```

Alternatively, open your browser's console and do:
```
let m = document.createElement( 'meta' );
m.setAttribute( 'property', 'og:image' );
m.content = 'https://example.com/image.jpg?a=b&c=d;'
m.outerHTML;
```
You'll see this output:
```
<meta property="og:image" content="https://example.com/image.jpg?a=b&amp;c=d;">
```

If something is broken, my guess is that it's because something somewhere else in the stack is not correctly parsing the HTML `<meta>` element.

**Comment by leerob** [comment, marked as answer]
"That's the behavior of `react-dom/server`. Here's a codesandbox using the snippet you're referencing.

https://q6o1b.csb.app"

https://twitter.com/rauchg/status/1374190846963113989

**Comment by phips28** [comment]
Is there still no real fix for that? I just came across this, where I added an og:image in nextjs 14.x (Metadata) 🙈

**Reply by phips28** [reply]
Oh, I now see this problem only occurs on localhost. When I deploy to Vercel, the URL is correct.
